public/app/features/dashboard/components/GenAI/GenAIButton.tsx

```diff
@@ -26,38 +28,89 @@ export interface GenAIButtonProps {
 export const GenAIButton = ({
   text = 'Auto-generate',
   loadingText = 'Generating',
+  toggleTipTitle = '',
   onClick: onClickProp,
   messages,
   onGenerate,
   temperature = 1,
 }: GenAIButtonProps) => {
   const styles = useStyles2(getStyles);
 
-  // TODO: Implement error handling (use error object from hook)
-  const { setMessages, reply, isGenerating, value, error } = useOpenAIStream(OPEN_AI_MODEL, temperature);
+  const [history, setHistory] = useState<string[]>([]);
+  const [shouldCloseHistory, setShouldCloseHistory] = useState(false);
+
+  const { setMessages, reply, value, error, streamStatus } = useOpenAIStream(OPEN_AI_MODEL, temperature);
+
+  const hasHistory = history.length > 0;
+  const isFirstGeneration = streamStatus === StreamStatus.GENERATING && !hasHistory;
+  const isButtonDisabled = isFirstGeneration || (value && !value.enabled && !error);
 
   const onClick = (e: React.MouseEvent<HTMLButtonElement>) => {
-    onClickProp?.(e);
-    setMessages(messages);
+    if (!hasHistory) {
+      onClickProp?.(e);
+      setMessages(messages);
+    }
   };
 
+  const updateHistory = useCallback(
```

Comment:
```suggestion
  const pushHistoryEntry = useCallback(
```
It was hard for me to understand what it was updating, what do you think about this? Like browser history 😛